### ISSUE TITLE
Keyboard layout switching unreliable

### ISSUE THREAD

**Issue by arkenoi** [QubesOS/qubes-issues#6517]
### Qubes OS version:

R4.1

### Affected component(s) or functionality:

GUI

---

### Steps to reproduce the behavior:

keyboard layout switcher is ru++grp:shifts_toggle

### Expected or desired behavior:

shift+shift consistently switches layouts in any VM

### Actual behavior:

layout switcher is reliable in dom0 but not in domU, sometimes it is not changed at all and gets out of sync with xfce layout indicator if there is one in place.

### General notes:

Apparently turning HT on makes things worse but I am not sure

---

### I am aware of the following related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues:

https://github.com/QubesOS/qubes-issues/issues/1396

**Comment by arkenoi** [comment]
Don't know if it helps to diagnose, but it happens if you switch and type with no pause. If you switch, wait and type after a second or two, it works properly.

**Comment by arkenoi** [comment]
double checked, no, HT does not affect it. but i have the impression than it was better while on 3.x

**Comment by DemiMarie** [comment]
> Don't know if it helps to diagnose, but it happens if you switch and type with no pause. If you switch, wait and type after a second or two, it works properly.

This sounds like a race condition.  I imagine that the keyboard layout switch is communicated via qrexec, which does not synchronize with the GUI daemon.  Ideally, keyboard layout events would be sent via the GUI protocol instead.

If I am correct, what happens is that you start typing before the qube you are typing into has been notified of the changed keyboard layout.  This should not be possible.

**Comment by hexagonrecursion** [comment]
Nitpick: the information is transferred not via qrexec, but via QubesDB
https://github.com/QubesOS/qubes-core-admin/blob/568908c8c628171a638e5b7a683aa55aeb33bc42/qubes/ext/gui.py#L92
```py
vm.untrusted_qdb.write('/keyboard-layout', vm.keyboard_layout)
```

**Comment by jamke** [comment]
I have the same issue, it happens in different qubes (at least template based usual ones).
**R4.1.1**.
Have not had it on **R4.0** (probably never), but had something similar on R3 releases.
I'm using the same keyboard as it was for R4.0, so it's not about the change of keyboard pressing style or timing change.

**Comment by jamke** [comment]
> This sounds like a race condition. I imagine that the keyboard layout switch is communicated via qrexec, which does not synchronize with the GUI daemon. Ideally, keyboard layout events would be sent via the GUI protocol instead.
> 
> If I am correct, what happens is that you start typing before the qube you are typing into has been notified of the changed keyboard layout. This should not be possible.

Any possible way to debug this situation? I have this layout-switch issue every day multiple times, and can help with diagnosis.
I had no such issue on R4.0, but the PC was different too.

**Potentially useful information about race condition:**
I use 2 layouts and toggle them with capslock to the first one and shift+capslock for the second (last) one (`shift_caps_switch` in terms of `setxkbmap`).
1) I noticed that switching to the first `us` layout ALWAYS work for me.
2) And in case I'm trying to switch to the second and it is broken - I can fix the situation like that:
- Press and release Capslock (switch to first)
- Wait for 1-2 seconds
- Press and release Shift+Capslock (switch to the second (aka last)).

It's a 100% reproducible fix but very annoying to do it everyday several times.

If I **don't wait 1-2 seconds** and try to shift+capslock - it **will never work** in this condition - the layout keeps being `us` (first).
So the issue **is about time waiting** and some race condition is quite possible.

**Comment by jamke** [comment]
Maybe useful information for this issue too: https://github.com/QubesOS/qubes-issues/issues/8035#issuecomment-1537514055

**Comment by andrewdavidwong** [comment]
@marmarek: Some users in [this forum thread](https://forum.qubes-os.org/t/how-qubes-installation-manages-keyboard-layout/20397/) have suggested that this issue is serious enough that the Qubes team may wish to consider it a blocker for 4.2. What do you think?

**Comment by jamke** [comment]
I think this issue can be considered as a combination of several problems: 
1. 8035 (fixed)
2. 8230 (fixed)
3. 8441

We are a couple of steps from making layout switching great again.

**Comment by jamke** [comment]
All these race condition stuff discussed above:
https://github.com/QubesOS/qubes-issues/issues/6517#issuecomment-833410602
https://github.com/QubesOS/qubes-issues/issues/6517#issuecomment-833972502
https://github.com/QubesOS/qubes-issues/issues/6517#issuecomment-1357345533

can be related to this double events and a time gap between them described and diagnosed in the last issue on the list: 
#8441.

**Comment by UndeadDevel** [comment]
For me waiting does not fix anything, but thankfully the bug doesn't happen too often. I'll add my "variant" of this problem:

I have three layouts set up, let's call them A, B, C (in that order);
- in dom0 they all always work
- A and B always work in domUs
- if a domU is started with C already selected, then C will always work (and A and B also)
- C occasionally will not work in domU, if A or B was selected when that domU started, even if I try to wait and switch again etc...it won't work until I restart that domU with C already set as global layout

**Comment by jamke** [comment]
@UndeadDevel it may be useful if you would be able to provide the output of `setxkbmap -print -query` from `dom0` and one of `domU` under all the cases (when works and does not work and in different expected layouts A, B, C)?

Are domUs templated-based qubes, based on `fedora-38` template?

**Comment by UndeadDevel** [comment]
@jamke I am using both fedora-38 and debian-12 templates as well as Whonix for my domUs...the bug seems to occur on all of them. I've also been able to make it more reproducible and I seem to have found a workaround for it:

**How to trigger the bug**: if it isn't already active in the domU currently being used then switching layouts seems to reliably trigger it...now that I've been paying more attention to it I've noticed it actually happens more often than I first thought.

**How to get rid of the bug**: switch layout to A while a window of the bugged domU is in focus, then switch focus to a window of a different VM (e.g. Alt+Tab or switching to different workspace), then switch the layout back to C and finally return focus to the bugged VM window and it should have that layout working now.

Here are the outputs from the command you suggested (note: A = us, B = B, C = de...I'm going to keep what B is to myself):

**dom0**
```
xkb_keymap {
        xkb_keycodes  { include "evdev+aliases(qwerty)" };
        xkb_types     { include "complete"      };
        xkb_compat    { include "complete"      };
        xkb_symbols   { include "pc+us+B:2+de:3+inet(evdev)+group(win_space_toggle)"    };
        xkb_geometry  { include "pc(pc105)"     };
};
rules:      evdev
model:      pc105
layout:     us,B,de
variant:    ,,
options:    grp:win_space_toggle
```

**domU bugged and C (de) selected**

```
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwertz)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+de+us:2+inet(evdev)+group(win_space_toggle)"};
	xkb_geometry  { include "pc(pc105)"	};
};
rules:      evdev
model:      pc105
layout:     de,us
variant:    ,
options:    grp:win_space_toggle
```
Note that switching to B in the bugged VM can be done successfully, even though it doesn't show up in the output.

**domU unbugged using above mentioned workaround and C (de) selected**
same as the prior output...

**domU bugged again by switching layout to A (us)**

```
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+inet(evdev)+group(win_space_toggle)"	};
	xkb_geometry  { include "pc(pc105)"	};
};
rules:      evdev
model:      pc105
layout:     us
options:    grp:win_space_toggle
```
(confirmed bugged after then switching to C, but I also can confirm that if I have C working and then use the switch three times to cycle from C to C it will be bugged again)


These tests were done on Whonix, but as I said the behavior is the same for normal debian-12 and also fedora-38, as well as fedora-38-minimal.

**Comment by marmarek** [comment]
```
options:    grp:win_space_toggle
```
You have this in domU too, right? This looks like the same issue as https://github.com/QubesOS/qubes-issues/issues/8035, and that fix should help your case too. Fixed package is in R4.2 repository, but you are on R4.1, right? I'll backport the fix in coming days.

**Comment by UndeadDevel** [comment]
> You have this in domU too, right?

Yes, that line is part of the output in all domUs it seems.

> you are on R4.1, right?

Yes.

> I'll backport the fix in coming days.

Awesome, thanks!

**Comment by alimirjamali** [comment]
PR Submitted.

Review priority: Medium (since it involves multiple reports/complains in Github and Forum. Also the other linked issue has major label).

Technical details are available in PR commit comments.

**Comment by alimirjamali** [comment]
@andrewdavidwong I believe this issue has to be reopened.

While the patch I submitted very recently improves the situation a lot. We still have the race condition between `qubes-db` event dispatcher/listener and GUI daemon/agent (but at least it is now reliable, just slow).

Also affects r4.3

> This sounds like a race condition. I imagine that the keyboard layout switch is communicated via qrexec, which does not synchronize with the GUI daemon. Ideally, keyboard layout events would be sent via the GUI protocol instead.

I am afraid the only liable solution might be the above. Before that, I will try to do some measurements.

**Comment by marmarek** [comment]
It is asynchronous, so it will be applied in VM with a slight delay, but in practice the delay should be rather small (well under a second). I guess if somebody type _instantly_ after switching the layout it may still use the old layout for a character or two, but I don't think that's a common case and worth spending time on.
But if the delay is longer, there is something to fix.

**Comment by alimirjamali** [comment]
> It is asynchronous, so it will be applied in VM with a slight delay, but in practice the delay should be rather small (well under a second). I guess if somebody type _instantly_ after switching the layout it may still use the old layout for a character or two, but I don't think that's a common case and worth spending time on. But if the delay is longer, there is something to fix.

Yes. But I my typing speed is 100+ characters per minute (maybe over two characters per second). And I definitely encounter this race condition where GUI agent/daemon are faster than `qubes-db` event dispatcher/listener. I have personally built the habit of waiting half a second between switching layout and starting to type with new layout. This is still annoying.

**Comment by jamke** [comment]
> @andrewdavidwong I believe this issue has to be reopened.

I agree. For me keyboard layout switch is still kind of broken in R4.2 as I still have issues with both delay and not-switching in case of using `shift_caps_switch`. 

Scripts in `dom0` often do not send keyboard layout change event when required in such case.

**Comment by marmarek** [comment]
> still kind of broken in R4.2

The fix has not been backported to R4.2 yet.
